message_sender/tasks.py

```diff
@@ -77,6 +79,44 @@ def run(self, metric_name, metric_value, session=None, **kwargs):
 fire_metric = FireMetric()
 
 
+class Concurrency_Limiter(object):
+    def __init__(self):
+        self.redis_server = redis.StrictRedis(host=settings.REDIS_HOST,
+                                              port=settings.REDIS_PORT,
+                                              db=settings.REDIS_DB)
+
+    def get_current_message_count(self, msg_type):
+        keys = self.redis_server.keys(pattern=msg_type + "_messages_at_*")
+        total = 0
+        # Sum the values in all the buckets to get the total
+        for key in keys:
+            total += int(self.redis_server.get(key))
+        return total
+
+    def incr_message_count(self, msg_type, delay):
+        # Buckets of 1min
+        key = msg_type + "_messages_at_" + time.strftime("%M")
+        value = self.redis_server.incr(key, 1)
+        if value == 1:
+            # Add 60 seconds to the expiry time so messages that start at the
+            # end of the minute still complete
+            self.redis_server.expire(key, delay + 60)
+
+    def decr_message_count(self, msg_type, delay):
+        # Buckets of 1min
+        key = msg_type + "_messages_at_" + time.strftime("%M")
+        self.redis_server.decr(key, 1)
```

Comment:
We need to check if the result is `<0`, and if it is, set the key to zero, so that we never have negative numbers for keys. (This case could happen if the cache was wiped before all the events for all the outgoing messages had come in).